### PR TITLE
Java JSON example

### PR DESCRIPTION
**Pull request by dominikl**
# What this PR does

Adds an example for using the JSON API with Java.

# Testing this PR

Setup the additional dependencies (see comment in `JSONClient.java`) and run it with the necessary parameters (see comment in `JSONClient.java`).

# Related reading

https://github.com/openmicroscopy/ome-documentation/pull/1627
(respectively: http://www.openmicroscopy.org/site/support/omero5.3-staging/developers/json-api.html )

### CODE REVIEW

**Comment by dominikl** [comment]
It basically works, but I still get `HTTP/1.1 405 METHOD NOT ALLOWED` when trying to update an image name (although now using PUT), @will-moore .

**Comment by will-moore** [comment]
Ah - We only support creation or update of "Project", "Dataset" or "Screen".
So you can't update an Image Name. Too many issues with unloaded collections etc with saving Images at present.
This is mentioned under the [Creating and Saving Objects](http://www.openmicroscopy.org/site/support/omero5.3-staging/developers/json-api.html#creating-and-saving-objects) header, but I guess we should reiterate it under the "Updating Objects" section.

The content of that ```405``` response should have said "Update of Images not supported"?

Also, I will fix the POST -> ```PUT``` for updating objects.

**Comment by will-moore** [comment]
Updates in https://github.com/openmicroscopy/ome-documentation/pull/1627/commits/f937f90744f8ce957b59c78ce778fac62fafdd0d

**Comment by dominikl** [comment]
Thanks @will-moore . I'll modify the example to change the name of a dataset instead!

**Comment by dominikl** [comment]
Looks like the CI runs everything in the examples directory. This will obviously not work with this example. Have to find out how to disable this example; if not possible, I'll have to comment out the main method.
Edit: As @jburel pointed out, it's the compilation which fails. I'll have a look if I can put the example somewhere else, or some how manage to get the dependencies in for the build.

**Comment by jburel** [comment]
the file is recompiled

excluding for now

--exclude

**Comment by waxenegger** [comment]
tried locally -in eclipse- and worked nicely. altered ever so slightly to query plates...

```
API versions:
{"version":"0","url:base":"http://web-dev-merge.openmicroscopy.org/api/v0/"}
URLs:
url:login : http://web-dev-merge.openmicroscopy.org/api/v0/login/
url:datasets : http://web-dev-merge.openmicroscopy.org/api/v0/m/datasets/
url:token : http://web-dev-merge.openmicroscopy.org/api/v0/token/
url:screens : http://web-dev-merge.openmicroscopy.org/api/v0/m/screens/
url:servers : http://web-dev-merge.openmicroscopy.org/api/v0/servers/
url:plates : http://web-dev-merge.openmicroscopy.org/api/v0/m/plates/
url:save : http://web-dev-merge.openmicroscopy.org/api/v0/m/save/
url:projects : http://web-dev-merge.openmicroscopy.org/api/v0/m/projects/
url:schema : http://www.openmicroscopy.org/Schemas/OME/2016-06
url:images : http://web-dev-merge.openmicroscopy.org/api/v0/m/images/
Servers:
eel-latest : 2
seabass : 4
eel-merge : 1
eel-integration : 3
Log in to server eel-merge:
Logged in: {"userName":"user-3","eventId":-1,"sessionUuid":"******","eventType":"User","userId":9,"isAdmin":false,"groupName":"read-only-1","sessionId": ,"memberOfGroups":[8,7,1],"leaderOfGroups":[8],"groupId":8}
Plates:
{"Rows":1,"Name":"","RowNamingConvention":"letter","omero:details":{"owner":{"UserName":"user-4","FirstName":"user-4","MiddleName":"","omero:details":{"@type":"TBD#Details","permissions":{"isUserWrite":true,"isWorldWrite":false,"canDelete":false,"isWorldRead":false,"perm":"rw----","canEdit":false,"canAnnotate":false,"isGroupAnnotate":false,"isGroupWrite":false,"canLink":false,"isUserRead":true,"@type":"TBD#Permissions","isGroupRead":false}},"Email":"","LastName":"user-4","@id":5,"@type":"http://www.openmicroscopy.org/Schemas/OME/2016-06#Experimenter"},"group":{"omero:details":{"@type":"TBD#Details","permissions":{"isUserWrite":true,"isWorldWrite":false,"canDelete":false,"isWorldRead":false,"perm":"rwr---","canEdit":false,"canAnnotate":false,"isGroupAnnotate":false,"isGroupWrite":false,"canLink":false,"isUserRead":true,"@type":"TBD#Permissions","isGroupRead":true}},"@id":8,"@type":"http://www.openmicroscopy.org/Schemas/OME/2016-06#ExperimenterGroup","Name":"read-only-1"},"@type":"TBD#Details","permissions":{"isUserWrite":true,"isWorldWrite":false,"canDelete":true,"isWorldRead":false,"perm":"rwr---","canEdit":true,"canAnnotate":true,"isGroupAnnotate":false,"isGroupWrite":false,"canLink":true,"isUserRead":true,"@type":"TBD#Permissions","isGroupRead":true}},"url:screens":"http://web-dev-merge.openmicroscopy.org/api/v0/m/plates/1951/screens/","ColumnNamingConvention":"number","url:plateacquisitions":"http://web-dev-merge.openmicroscopy.org/api/v0/m/plates/1951/plateacquisitions/","url:wells":"http://web-dev-merge.openmicroscopy.org/api/v0/m/plates/1951/wells/","@id":1951,"@type":"http://www.openmicroscopy.org/Schemas/OME/2016-06#Plate","Columns":2,"url:plate":"http://web-dev-merge.openmicroscopy.org/api/v0/m/plates/1951/"}
{"Rows":3,"Description":"Plate 1 description.","omero:details":{"owner":{"UserName":"user-4","FirstName":"user-4","MiddleName":"","omero:details":{"@type":"TBD#Details","permissions":{"isUserWrite":true,"isWorldWrite":false,"canDelete":false,"isWorldRead":false,"perm":"rw----","canEdit":false,"canAnnotate":false,"isGroupAnnotate":false,"isGroupWrite":false,"canLink":false,"isUserRead":true,"@type":"TBD#Permissions","isGroupRead":false}},"Email":"","LastName":"user-4","@id":5,"@type":"http://www.openmicroscopy.org/Schemas/OME/2016-06#Experimenter"},"group":{"omero:details":{"@type":"TBD#Details","permissions":{"isUserWrite":true,"isWorldWrite":false,"canDelete":false,"isWorldRead":false,"perm":"rwr---","canEdit":false,"canAnnotate":false,"isGroupAnnotate":false,"isGroupWrite":false,"canLink":false,"isUserRead":true,"@type":"TBD#Permissions","isGroupRead":true}},"@id":8,"@type":"http://www.openmicroscopy.org/Schemas/OME/2016-06#ExperimenterGroup","Name":"read-only-1"},"@type":"TBD#Details","permissions":{"isUserWrite":true,"isWorldWrite":false,"canDelete":true,"isWorldRead":false,"perm":"rwr---","canEdit":true,"canAnnotate":true,"isGroupAnnotate":false,"isGroupWrite":false,"canLink":true,"isUserRead":true,"@type":"TBD#Permissions","isGroupRead":true}},"url:screens":"http://web-dev-merge.openmicroscopy.org/api/v0/m/plates/2084/screens/","url:plateacquisitions":"http://web-dev-merge.openmicroscopy.org/api/v0/m/plates/2084/plateacquisitions/","url:plate":"http://web-dev-merge.openmicroscopy.org/api/v0/m/plates/2084/","url:wells":"http://web-dev-merge.openmicroscopy.org/api/v0/m/plates/2084/wells/","@id":2084,"@type":"http://www.openmicroscopy.org/Schemas/OME/2016-06#Plate","Columns":3,"Name":""}
{"Rows":8,"Name":"001009001.flex","RowNamingConvention":"letter","omero:details":{"owner":{"UserName":"user-4","FirstName":"user-4","MiddleName":"","omero:details":{"@type":"TBD#Details","permissions":{"isUserWrite":true,"isWorldWrite":false,"canDelete":false,"isWorldRead":false,"perm":"rw----","canEdit":false,"canAnnotate":false,"isGroupAnnotate":false,"isGroupWrite":false,"canLink":false,"isUserRead":true,"@type":"TBD#Permissions","isGroupRead":false}},"Email":"","LastName":"user-4","@id":5,"@type":"http://www.openmicroscopy.org/Schemas/OME/2016-06#Experimenter"},"group":{"omero:details":{"@type":"TBD#Details","permissions":{"isUserWrite":true,"isWorldWrite":false,"canDelete":false,"isWorldRead":false,"perm":"rwr---","canEdit":false,"canAnnotate":false,"isGroupAnnotate":false,"isGroupWrite":false,"canLink":false,"isUserRead":true,"@type":"TBD#Permissions","isGroupRead":true}},"@id":8,"@type":"http://www.openmicroscopy.org/Schemas/OME/2016-06#ExperimenterGroup","Name":"read-only-1"},"@type":"TBD#Details","permissions":{"isUserWrite":true,"isWorldWrite":false,"canDelete":true,"isWorldRead":false,"perm":"rwr---","canEdit":true,"canAnnotate":true,"isGroupAnnotate":false,"isGroupWrite":false,"canLink":true,"isUserRead":true,"@type":"TBD#Permissions","isGroupRead":true}},"ExternalIdentifier":"JL_120516_DynTest_Ctrl","url:screens":"http://web-dev-merge.openmicroscopy.org/api/v0/m/plates/1852/screens/","ColumnNamingConvention":"number","url:plateacquisitions":"http://web-dev-merge.openmicroscopy.org/api/v0/m/plates/1852/plateacquisitions/","url:wells":"http://web-dev-merge.openmicroscopy.org/api/v0/m/plates/1852/wells/","@id":1852,"@type":"http://www.openmicroscopy.org/Schemas/OME/2016-06#Plate","Columns":12,"url:plate":"http://web-dev-merge.openmicroscopy.org/api/v0/m/plates/1852/"}

....
```

it might be useful to add a code remark to the dependencies that a json api implementation is needed on top of the json javax interfaces but then again people would probably find out quickly anyhow or they might even use different http and json libraries.

either way I think this example gives developers a very good idea how to interact with the api using java.

**Comment by jburel** [comment]
one option will be to add the dependencies to the omero_clients.jar since it is added to the classpath when the file is compiled in CI job

**Comment by dominikl** [comment]
Added the dependencies and removed 'exclude' comment, that should hopefully fix the compile errors (but the example still is not run by default).

**Comment by jburel** [comment]
@dominikl what will you need to run it by default?
the job has been modified so the ``JsonLogin`` example in python can be run

**Comment by dominikl** [comment]
Should be possible to run it automatically. Will try to add to Setup.java later.

**Comment by mtbc** [comment]
Suggestion: try again in CI after wiping out that node's `.m2/repository/commons-logging/` directory.

**Comment by dominikl** [comment]
If I put the URL into the ivy dependency and set the force flag, it works for me. If not, I noticed that jar files were either not downloaded or they were missing some class files.
The last two commits should hopefully fix the compile errors and also run the JSON example automatically (added it to Setup.java).
But I keep the exclude flag for now, to not cause unnecessary delays for the 5.3.0 testing builds.

**Comment by dominikl** [comment]
Removed the exclude comment. Just add it again, if there are still problems with the build @jburel .

File: components/tools/OmeroJava/ivy.xml [hunk 1]
```diff
@@ -41,9 +41,18 @@
     <dependency org="ome" name="formats-gpl" rev="${versions.bioformats}" transitive="true">
         <exclude org="org.slf4j"/>
     </dependency>
-    <dependency org="org.glassfish" name="javax.json" rev="${versions.json}" transitive="true"/>
-    <dependency org="org.apache.httpcomponents" name="httpclient" rev="${versions.httpclient}" transitive="true"/>
-    <dependency org="org.apache.httpcomponents" name="httpmime" rev="${versions.httpclient}" transitive="true"/>
+    <dependency org="org.glassfish" name="javax.json" rev="1.0.4" force="true" transitive="true">
```
Comment (jburel) [inline components/tools/OmeroJava/ivy.xml]: why not retrieving the version from ``omero.properties``?
and keep ``force``

**Comment by jburel** [comment]
https://ci.openmicroscopy.org/job/OMERO-DEV-merge-integration/604/console is red

**Comment by jburel** [comment]
Let' hold off on that one.
The example works but some adjustments are required to integrate it in CI
Similar to what we did for https://github.com/openmicroscopy/openmicroscopy/pull/5116
let's close it for now but still exposes the example.

But could you open a PR against @will-moore PR https://github.com/openmicroscopy/ome-documentation/pull/1627
and add the example under https://github.com/openmicroscopy/ome-documentation/tree/develop/omero/examples/
and add link in ``omero/developers/json-api.txt``

cc @will-moore

**Comment by dominikl** [comment]
👍 I'll add the example to the documentation.

**Pull request closed without merging**